util/sym-mode_test.go

```diff
@@ -0,0 +1,58 @@
+// Mgmt
+// Copyright (C) 2013-2019+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package util_test
+
+import (
+	"os"
+	"testing"
+
+	"github.com/purpleidea/mgmt/util"
+)
+
+func TestSymbolicMode(t *testing.T) {
+	def := os.FileMode(0644) | os.ModeSetgid
+	symModeTests := []struct {
+		input  string
+		expect os.FileMode
+	}{
+		{"ug=rwx", 0774 | os.ModeSetgid},
+		{"ug=srwx", 0774 | os.ModeSetgid | os.ModeSetuid},
+		{"ug=trwx", 0774 | os.ModeSetgid | os.ModeSticky},
+		{"o=rx", 0645 | os.ModeSetgid},
+		{"ug=srwx", 0774 | os.ModeSetgid | os.ModeSetuid},
+		{"o+rwx", 0647 | os.ModeSetgid},
+		{"u+x", 0744 | os.ModeSetgid},
+		{"u+x", 0744 | os.ModeSetgid},
+		{"u+s", 0644 | os.ModeSetgid | os.ModeSetuid},
+		{"u+t", 0644 | os.ModeSetgid | os.ModeSticky},
+		{"o-rwx", 0640 | os.ModeSetgid},
+		{"u-w", 0444 | os.ModeSetgid},
+		{"g-s", 0644},
+		{"u-t", 0644 | os.ModeSetgid},
+	}
+
+	for _, test := range symModeTests {
+		t.Run(test.input, func(t *testing.T) {
```

Comment:
Needs to be race free... See other tests to avoid this. https://golang.org/doc/articles/race_detector.html#Race_on_loop_counter

Reply:
Sorry I thought range with `t.Run` was race free because each run has its own struct from the slice. Does a copy need to be made before `t.Run()`?

Reply:
Honestly I have to dig into the code again and check, but at one time I'm fairly sure it wasn't race free. So I did this: https://github.com/purpleidea/mgmt/blob/2ea492c96512a0802c98c9a06984457e88d5f2a4/lang/interpret_test.go#L431